rdbms/src/main/java/org/spine3/server/storage/jdbc/entity/visibility/VisibilityHandler.java

```diff
@@ -0,0 +1,164 @@
+
+    /**
+     * Creates a new instance of the {@code VisibilityHandler}
+     *
+     * @param queryFactory a factory for generating the queries to the table in which the visibility
+     *                     is stored
+     */
+    public VisibilityHandler(VisibilityHandlingStorageQueryFactory<I> queryFactory) {
+        this.queryFactory = checkNotNull(queryFactory);
+    }
+
+    /**
+     * Creates the table if it doesn't exist yet.
+     *
+     * <p>SQL equivalent of calling this method is:
+     * <pre>
+     *   <code>
+     *     CREATE TABLE IF NOT EXISTS %table-name% (
+     *         id VARCHAR(512),
+     *         archived BOOLEAN,
+     *         deleted BOOLEAN
+     *     );
+     *   </code>
+     * </pre>
+     *
+     * <p>Common usage is to call this method right after the constructor, but this is left
+     * on behalf of the class user.
+     */
+    public void initialize() {
+        queryFactory.newCreateVisibilityTableQuery()
+                    .execute();
+    }
+
+    /**
+     * Reads the {@linkplain Visibility} from the database.
+     *
+     * @param id ID of the record
+     * @return {@code Optional.absent()} if the {@linkplain Visibility} was not found or was
+     * {@linkplain org.spine3.validate.Validate#isDefault(Message) default}, {@code Optional.of)}
+     * the record otherwise
+     */
+    public Optional<Visibility> readVisibility(I id) {
+        final SelectVisibilityQuery query = queryFactory.newSelectVisibilityQuery(id);
+        final Visibility visibility = query.execute();
+        final boolean absent = visibility == null || isDefault(visibility);
+        if (absent) {
+            return Optional.absent();
+        }
+        return Optional.of(visibility);
+    }
+
+    /**
+     * Inserts a new {@linkplain Visibility} into the database or updates an existing one.
+     *
+     * @param id         ID of the record to store
+     * @param visibility the {@linkplain Visibility} to store
+     */
+    public void writeVisibility(I id, Visibility visibility) {
+        final Optional<Visibility> current = readVisibility(id);
+        final boolean exists = current.isPresent();
+        writeVisibility(id, visibility, exists);
+    }
+
+    private void writeVisibility(I id, Visibility status, boolean updateExisting) {
+        if (updateExisting) {
+            updateVisibility(id, status);
+        } else {
+            insertVisibility(id, status);
+        }
+    }
+
+    private void updateVisibility(I id, Visibility status) {
+        queryFactory.newUpdateVisibilityQuery(id, status)
+                    .execute();
+    }
+
+    private void insertVisibility(I id, Visibility status) {
+        queryFactory.newInsertVisibilityQuery(id, status)
+                    .execute();
+    }
+
+    /**
+     * Creates a new record as archived or updates an existing one to become archived.
+     *
+     * <p>Effectively performs no action if the record exists and is archived.
```

Comment:
`Has no effect if the record exists and has been marked archived previously.`

Same below.